### DISCUSSION TITLE
Unable to Fetch Page Data in Next13

### DISCUSSION THREAD

**Discussion by harblaith7** [vercel/next.js#44724] [Help]
When I try to fetch data from within my `page.tsx` file, I am periodically shown this error

<img width="1250" alt="Screen Shot 2023-01-09 at 9 33 38 AM" src="https://user-images.githubusercontent.com/35265876/211332568-de432165-5fcd-4e91-b207-458aadff8742.png">

Usually, I restart the server and it goes away, but it comes back every now and then. This is how my code looks. I copied what was recommended in the [documentation](https://beta.nextjs.org/docs/data-fetching/fetching)

```react
const fetchRestaurant = async (slug: string): Promise<Restaurant> => {
  const res = await fetch(`http://localhost:3000/api/restaurant/${slug}`);
  return res.json();
};

export default async function RestaurantPage({
  params,
}: {
  params: { name: string };
}) {
  const restaurant = await fetchRestaurant(params.name);
  return (
    <main>
      
    </main>
  );
}
```

Am I doing something incorrectly, or is this a problem is Next13 right now?

**Comment by icyJoseph** [comment]
Hi,

In the link you provide there's no instances of calling localhost api routes, like:

```js
const res = await fetch(`http://localhost:3000/api/restaurant/${slug}`);
```

You should not rely on the presence of an API handler while you are within the server itself. There's a few reasons going from the technical impossibility to whether or not this is optimal.

The next js docs state, https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props#getserversideprops-or-api-routes:

> It can be tempting to reach for an [API Route](https://nextjs.org/docs/api-routes/introduction) when you want to fetch data from the server, then call that API route from getServerSideProps. This is an unnecessary and inefficient approach, as it will cause an extra request to be made due to both getServerSideProps and API Routes running on the server.

The above applies to Server Components, and GetStaticProps as well - albeit with some caveats.

During build time, for GetStaticProps, or Server Component pre-rendering, there's no API routes, the server doesn't exist. It's just workers rendering React into HTML.

The solution is to not depend on API routes, but rather parametrize the request to 3rd parties, for example you calculate restaurants as a function of `slug` - So you can write  a function that takes a slug and directly gets your restaurants from whatever service you use:

```ts
const fetchRestaurant = async (slug: string): Promise<Restaurant> => {
  const res = await fetch(`http://my-restaurant-api.com/${slug}`);
  return res.json();
};
```

Now you can use this function in Server Components, and for example API routes:

```tsx
export default async function RestaurantPage({
  params,
}: {
  params: { name: string };
}) {
  const restaurant = await fetchRestaurant(params.name);
  return (
    <main>
      
    </main>
  );
}
```
or

```ts
import {NextApiRequest, NextApiResponse} from 'next';
// pseudo-code
export const handler = async (req: NextApiRequest, res: NextApiResponse) => {
    try {
        const {slug} = req.query
        const data = await fetchRestaurants(slug);
        return res.json(data).end();
    } catch (err) {
       return res.status(500).end();
    }
};

export default handler;
```

And there you go.

**Reply by mike14747** [reply]
Just wondering... why does everyone seem to use fetch in Server Components, rather than just directly calling the same serverless functions the api route utilizes?

```js
import { getUnprotectedData } from '../../lib/api/index';

async function getData() {
    return await getUnprotectedData();
}

export default async function Page() {
    const data = await getData();

    return (
        <>
            {/* map the data */}
        </>
    );
}
```

**Reply by icyJoseph** [reply]
Right, in Nuxt for instance, it is possible to do that, https://nuxt.com/docs/guide/concepts/server-engine#direct-api-calls:

> Nitro allows 'direct' calling of routes via the globally-available $fetch helper. This will make an API call to the server if run on the browser, but will directly call the relevant function if run on the server, saving an additional API call.

In Next.js, to call a the API route handler directly, you need to provide, the NextRequest, and NextResponse objects (classes). Technically, nothing is preventing us from instantiating those two and passing them to the handler. However, the instantiation is the actual problem, because it is not public, so we have to "mock" them. Next.js internally does create mock objects for certain steps of the build process. These are private APIs though, so we'd need to craft our own, and that's rather difficult to maintain. At least during the 11->13 evolution.

We in fact do this when writing certain Jest tests, I have NextRequest and NextResponse mocks. Yes we do favour actual Network mocking via MSW, or `next-test-api-route-handler`, or even `ts-mockito`, but sometimes you just want a simple-self-made mock. The effort  of maintaining these up to date has been non-zero, and that's already enough to discourage me from mocking them on production code.

The alternative to refactor my code so that I invoke a function that's based off fetch and other parameters, and that can be called from Server Components, GSSP, or API routes, is far better, and it helps keeping code clean and structured. One could even parametrise fetch, as an agent, and then the entire function is independent of the environment where it has to run.

Yet another alternative, would be to make your API handlers polymorphic, and handle the case where NextRequest and NextResponse, are undefined, and instead expect other kind of values... again, technically, most things are possible, the question is whether or not, they are the right thing to do.

**Reply by AnirudhSinghBhadauria** [reply]
one of the best things about next13 is that we can make component function async itself, providing us superpowers to getData inside of server component itself. There is no point of fetching data from local api routes inside of server components if you are using something like prisma, you can directly call prisma client inside of server component and get your data. Using api routes will add an additional layer of code blocking only and thats not very healthy according to NEXT team.
